haystack/agents/base.py

```diff
@@ -0,0 +1,298 @@
+        """
+        self.tools[tool.name] = tool
+
+    def has_tool(self, tool_name: str):
+        """
+        Check whether the Agent has a Tool registered under the provided tool name.
+
+        :param tool_name: The name of the Tool for which to check whether the Agent has it.
+        """
+        return tool_name in self.tools
+
+    def run(
+        self, query: str, max_iterations: Optional[int] = None, params: Optional[dict] = None
+    ) -> Dict[str, Union[str, List[Answer]]]:
+        """
+        Runs the Agent given a query and optional parameters to pass on to the tools used. The result is in the
+        same format as a pipeline's result: a dictionary with a key `answers` containing a List of Answers
+
+        :param query: The search query.
+        :param max_iterations: The number of times the Agent can run a tool plus then once infer it knows the final answer.
+            Set at least to 2 so that the Agent can run one tool and then infer it knows the final answer. Default 5.
+        :param params: A dictionary of parameters that you want to pass to those tools that are pipelines.
+                       To pass a parameter to all nodes in those pipelines, use: `{"top_k": 10}`.
+                       To pass a parameter to targeted nodes in those pipelines, use:
+                        `{"Retriever": {"top_k": 10}, "Reader": {"top_k": 3}}`.
+                        Parameters can only be passed to tools that are pipelines but not nodes.
+        """
+        if len(self.tools) == 0:
+            raise AgentError(
+                "Agents without tools cannot be run. Add at least one tool using `add_tool()` or set the parameter `tools` when initializing the Agent."
+            )
+        if max_iterations is None:
+            max_iterations = self.max_iterations
+
+        transcript = ""
+        # Generate string representations of the tools to be passed to the PromptTemplate of the Agent
+        tool_names = ", ".join(self.tools.keys())
+        tool_names_with_descriptions = "\n".join([f"{tool.name}: {tool.description}" for tool in self.tools.values()])
+
+        # Generate a thought with a plan what to do, choose a tool, generate input for it, and run it.
+        # Repeat this until the final answer is found or the maximum number of iterations is reached.
+        for _ in range(max_iterations):
+            pred = self.prompt_node.prompt(
+                prompt_template=self.prompt_template,
+                query=query,
+                tool_names=tool_names,
+                tool_names_with_descriptions=tool_names_with_descriptions,
+                generated_text=transcript,
+            )[0]
+
+            # Try to extract the final answer or a tool name and tool input from the generated text
+            final_answer = self._extract_final_answer(pred=pred)
+            if final_answer is not None:
```

Comment:
Python perfectionists ask for `if final_answer:`

Reply:
If `final_answer` is `""` that evaluates to `False` so I would keep that as is.